<a href="https://colab.research.google.com/github/FaQ2108/FacundoMauro-Academics-Data-Projects/blob/main/IntAlMachLearn3_MauroFacundo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
import seaborn as sns

a) Debemos describir brevemente la cantidad de valores faltantes para
cada una de las variables de la base de datos.

In [2]:
dataset = pd.read_csv('train.csv')
dataset.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
#Ahora vamos a verificar cuantos registros faltan en este dataset:
dataset.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

b) Completemos aquellas variables que se encuentran faltantes para las
bases de train (“Age”, ”fare”, “Cabin”) y test ( “Age”, “Cabin”,
“embarked”).

In [4]:
#Completamos los registros faltantes en la columna 'Age' con la edad promedio del conjunto de edades registradas:
dataset['Age'].fillna(dataset['Age'].mean(), inplace=True)
#Verificamos la columna 'Age' para ver que efectivamente no existan registros faltantes:
dataset.isnull().sum()

<ipython-input-4-e0b541fc4215>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dataset['Age'].fillna(dataset['Age'].mean(), inplace=True)


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [5]:
#Ahora quiero saber cuales son los valores que puede tomar la columna 'Cabin'.
dataset['Cabin'].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [6]:
#Procedo a reemplazar los valores faltantes de la columna 'Cabin' con la palabra Unknow(Desconocido):
dataset['Cabin'].fillna('Unknown', inplace=True)
#Verificamos la columna 'Cabin' para ver que efectivamente no existan registros faltantes:
dataset.isnull().sum()

<ipython-input-6-b536c1bafd1c>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dataset['Cabin'].fillna('Unknown', inplace=True)


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       2
dtype: int64

In [7]:
#Finalmente verificamos los valores que puede tomar la columna 'Embarked':
dataset['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [8]:
#Reemplazamos los registros faltantes de la columna 'Embarked' con el valor que mas se repite entre los valores de la columna:
dataset['Embarked'].fillna(dataset['Embarked'].mode()[0], inplace=True)
#Verificamos la columna 'Embarked' para ver que efectivamente no existan registros faltantes:
dataset.isnull().sum()

<ipython-input-8-ba09ed1bc781>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dataset['Embarked'].fillna(dataset['Embarked'].mode()[0], inplace=True)


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [9]:
# Procedemos a cambiar los valores de las columnas 'Sex' y 'Embarked' a valores numéricos:
dataset['Sex'] = dataset['Sex'].map({'male': 0, 'female': 1})
dataset['Embarked'] = dataset['Embarked'].map({'S': 0, 'C': 1, 'Q': 2})

In [10]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    int64  
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    object 
 11  Embarked     891 non-null    int64  
dtypes: float64(2), int64(7), object(3)
memory usage: 83.7+ KB


c) Ahora, debemos ajustar nuestro primer modelo de árbol de decisión.

In [11]:
#Definimos las variables predictoras (X) y la variable objetivo (y):
X = dataset[['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']]
y = dataset['Survived']

In [12]:
#Dividimos los datos en conjuntos de entrenamiento y prueba:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [13]:
#Creamos un modelo de árbol de decisión:
tree = DecisionTreeClassifier(random_state=42)

In [14]:
#Entrenamos el modelo con los datos de entrenamiento:
tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [15]:
# Cargamos el conjunto de pruebas
test_data = pd.read_csv('test.csv')

In [16]:
# Realizamos las mismas transformaciones en el conjunto de pruebas que en el conjunto de entrenamiento
test_data['Age'].fillna(test_data['Age'].mean(), inplace=True)
test_data['Cabin'].fillna('Unknown', inplace=True)
test_data['Embarked'].fillna(test_data['Embarked'].mode()[0], inplace=True)
test_data['Fare'].fillna(test_data['Fare'].mean(), inplace=True) # completar Fare en test

test_data['Sex'] = test_data['Sex'].map({'male': 0, 'female': 1})
test_data['Embarked'] = test_data['Embarked'].map({'S': 0, 'C': 1, 'Q': 2})

test_data.isnull().sum()

<ipython-input-16-134a48161946>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  test_data['Age'].fillna(test_data['Age'].mean(), inplace=True)
<ipython-input-16-134a48161946>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=Tr

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [17]:
# Seleccionamos las mismas características que en el entrenamiento
X_test_final = test_data[['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']]

In [18]:
# Realizamos las predicciones
y_pred_test = tree.predict(X_test_final)

d) Finalmente, podemos entrenar y determinar Accuracy, F1-Score,
Recall.

In [19]:
#Evaluamos el rendimiento del modelo:

from sklearn.metrics import accuracy_score, f1_score, recall_score

# Realizamos predicciones en el conjunto de prueba
y_pred = tree.predict(X_test)

# Calculamos las métricas de rendimiento
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print(f"F1-Score: {f1}")
print(f"Recall: {recall}")

Accuracy: 0.776536312849162
F1-Score: 0.726027397260274
Recall: 0.7162162162162162


* **Accuracy (Precisión):** 0.7765. El modelo predice correctamente aproximadamente el 77.65% de las veces.

* **F1-Score:** 0.7260. Esta métrica es el promedio armónico de la precisión y el recall, y es útil cuando tienes un balance entre falsos positivos y falsos negativos. Un F1-Score de 0.7260 indica un buen equilibrio entre precisión y recall.

* **Recall (Sensibilidad):** 0.7162. Esta métrica mide la capacidad del modelo para identificar correctamente las instancias positivas. Un recall de 0.7162 significa que el modelo identifica correctamente el 71.62% de las instancias positivas.

## Conclusion:

El modelo de árbol de decisión ha demostrado un rendimiento sólido en la tarea de clasificación. Con una precisión del 77.65%, el modelo predice correctamente la mayoría de las instancias. El F1-Score de 0.7260 indica un buen equilibrio entre precisión y recall, lo que sugiere que el modelo maneja bien tanto los falsos positivos como los falsos negativos. Además, el recall del 71.62% muestra que el modelo es eficaz en la identificación de instancias positivas.